src/k2/persistence/plog_service/PlogServer.cpp

```diff
@@ -0,0 +1,161 @@
+#include <k2/transport/Payload.h>
+#include <k2/transport/Status.h>
+#include <k2/dto/Plog.h>
+#include <k2/common/Common.h>
+#include <k2/config/Config.h>
+#include <k2/cpo/client/CPOClient.h>
+#include <k2/transport/BaseTypes.h>
+#include <k2/transport/TXEndpoint.h>
+#include <k2/appbase/AppEssentials.h>
+#include <k2/appbase/Appbase.h>
+
+
+namespace k2 {
+
+PlogServer::PlogServer() {
+    K2INFO("ctor");
+}
+
+PlogServer::~PlogServer() {
+    K2INFO("dtor");
+}
+
+seastar::future<> PlogServer::gracefulStop() {
+    K2INFO("stop");
+    _plogMap.clear();
+    return seastar::make_ready_future<>();
+}
+
+seastar::future<> PlogServer::start() {
+    K2INFO("Registering message handlers");
+    RPC().registerRPCObserver<dto::PlogCreateRequest, dto::PlogCreateResponse>(dto::Verbs::K23SI_PERSISTENT_CREATE, [this](dto::PlogCreateRequest&& request) {
+        return _handleCreate(std::move(request));
+    });
+
+    RPC().registerRPCObserver<dto::PlogAppendRequest, dto::PlogAppendResponse>(dto::Verbs::K23SI_PERSISTENT_APPEND, [this](dto::PlogAppendRequest&& request) {
+        return _handleAppend(std::move(request));
+    });
+
+    RPC().registerRPCObserver<dto::PlogReadRequest, dto::PlogReadResponse>(dto::Verbs::K23SI_PERSISTENT_READ, [this](dto::PlogReadRequest&& request) {
+        return _handleRead(std::move(request));
+    });
+
+    RPC().registerRPCObserver<dto::PlogSealRequest, dto::PlogSealResponse>(dto::Verbs::K23SI_PERSISTENT_SEAL, [this](dto::PlogSealRequest&& request) {
+        return _handleSeal(std::move(request));
+    });
+    _plogMap.clear();
+
+    return seastar::make_ready_future<>();
+}
+
+seastar::future<std::tuple<Status, dto::PlogCreateResponse>>
+PlogServer::_handleCreate(dto::PlogCreateRequest&& request){
+    K2DEBUG("Received create request for " << request.plogId);
+    auto iter = _plogMap.find(request.plogId);
+    if (iter != _plogMap.end()) {
+        return RPCResponse(Statuses::S530_PlogId_is_duplicated("plog id already exists"), dto::PlogCreateResponse());
+    }
+    _plogMap.insert(std::pair<String,PlogPage >(std::move(request.plogId), PlogPage()));
+    return RPCResponse(Statuses::S201_Created("plog created"), dto::PlogCreateResponse());
+};
+
+seastar::future<std::tuple<Status, dto::PlogAppendResponse>>
+PlogServer::_handleAppend(dto::PlogAppendRequest&& request){
+    K2DEBUG("Received append request for " << request.plogId << " with size" << request.payload.getSize() << " and offset " << request.offset);
+    auto iter = _plogMap.find(request.plogId);
+    if (iter == _plogMap.end()) {
+        return RPCResponse(Statuses::S404_Not_Found("plog does not exist"), dto::PlogAppendResponse());
+    }
+    if (iter->second.sealed){
+         return RPCResponse(Statuses::S409_Conflict("plog is sealed"), dto::PlogAppendResponse());
+    }
+    if (iter->second.offset != request.offset){
+        return RPCResponse(Statuses::S400_Bad_Request("offset inconsistent"), dto::PlogAppendResponse());

```

Comment:
you should try to return different codes if you want the user to handle them differently. you return 400_bad_request in both cases here.

In particular, a mismatch in offset i think is a good candidate for 403_Forbidden, while 413 Request Entity Too Large i think works better for the case where the user wants to append too much data

Reply:
fixed